external-crates/move/crates/move-compiler/src/naming/translate.rs

```diff
@@ -456,90 +471,184 @@ impl<'env> Context<'env> {
+            let block_type = *block_type;
+            if block_type.is_acceptable_usage(usage) {
+                Some(*label)
             } else {
-                let msg = format!(
-                    "Invalid usage of '{}' with a {} block label",
-                    verb, block_type
-                );
+                let msg = format!("Invalid usage of '{usage}' with a {block_type} block label",);
                 let mut diag = diag!(NameResolution::InvalidLabel, (loc, msg));
-                match expected_block_type {
-                    NominalBlockType::Loop => {
-                        diag.add_note("Loop labels may only be used with 'break' and 'continue', not 'return'");
+                diag.add_note(match block_type {
+                    NominalBlockType::Loop(_) => {
+                        "Loop labels may only be used with 'break' and 'continue', \
+                        not 'return'"
                     }
                     NominalBlockType::Block => {
-                        diag.add_note("Named block labels may only be used with 'return', not 'break' or 'continue'.");
+                        "Named block labels may only be used with 'return', \
+                        not 'break' or 'continue'."
                     }
-                }
+                    NominalBlockType::LambdaReturn | NominalBlockType::LambdaLoopCapture => {
+                        "Lambda block labels may only be used with 'return' or 'break', \
+                        not 'continue'."
+                    }
+                });
                 self.env.add_diag(diag);
                 None
             }
         } else {
-            let msg = format!("Invalid {}. Unbound label '{}", verb, name);
+            let msg = format!("Invalid {usage}. Unbound label '{name}");
             self.env
-                .add_diag(diag!(NameResolution::UnboundVariable, (loc, msg)));
+                .add_diag(diag!(NameResolution::UnboundLabel, (loc, msg)));
             None
         }
     }
 
-    fn exit_nominal_block(&mut self, loc: Loc) -> BlockLabel {
-        let (name, id, _) = self.nominal_blocks.pop().unwrap();
-        let nvar_ = N::Var_ { name, id, color: 0 };
-        BlockLabel(sp(loc, nvar_))
+    fn exit_nominal_block(&mut self) -> (BlockLabel, NominalBlockType) {
+        let (_name, label, name_type) = self.nominal_blocks.pop().unwrap();
+        (label, name_type)
+    }
+}
+
+fn block_label(loc: Loc, name: Option<Symbol>, id: u16) -> BlockLabel {
+    let is_implicit = name.is_none();
+    let name = name.unwrap_or(BlockLabel::IMPLICIT_LABEL_SYMBOL);
+    let var_ = N::Var_ { name, id, color: 0 };
+    let label = sp(loc, var_);
+    BlockLabel { label, is_implicit }
+}
+
+impl NominalBlockType {
+    // loops can have break or continue
+    // blocks can have return
+    // lambdas can have return or break
+    fn is_acceptable_usage(self, usage: NominalBlockUsage) -> bool {
```

Comment:
```suggestion
    fn is_acceptable_usage(&self, usage: &NominalBlockUsage) -> bool {
```

Reply:
Why? The size of the reference is definitely going to be larger (and I'm not sure if Rust will optimize them away)